---
title: "Kilifish PGLS"
author: "Joe McGirr"
date: "12/2/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages<-function(x){
  x<-as.character(match.call()[[2]])
  if (!require(x,character.only=TRUE)){
    install.packages(pkgs=x,repos="http://cran.r-project.org")
    require(x,character.only=TRUE)
  }
}

packages(MASS)
packages(ggplot2)
packages(gtools)
packages(pheatmap)
packages(cowplot)
packages(RColorBrewer)
packages(dplyr)
packages(tidyr)
packages(knitr)
packages(ggrepel)
packages(DESeq2)
packages(limma)
packages(edgeR)
packages(gplots)
packages(lattice)
packages(vsn)
packages(biomaRt)
packages(kableExtra)
packages(pheatmap)
packages(SummarizedExperiment)
packages(emmeans)
packages(data.table)
packages(ggpubr)

source("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/raw/main/DE/PGLS/fun_MeanVar_all.R")

# color-blind friendly 
# Wong, B. Points of view: Color blindness. Nat Methods (2011).
bla <- "#000000"
blu <- "#0072b2"
grb <- "#56b4e9"
lir <- "#cc79a7"
gre <- "#009e73"
red <- "#d55e00"
org <- "#e69f00"
yel <- "#f0e442"
gry<-  '#BBBBBB'

```

## Input: Salmon quantification counts. Output: library size normalized counts.
Same pipeline as Lisa's. Skip below to read in normalized counts.

```{r loadfiles, results='hide', include=FALSE, warning=FALSE, eval = FALSE}

counts_design <- read.csv("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/raw/main/expression_tables/Ensembl_species_counts_designfactors.csv",stringsAsFactors = FALSE)




# I removed the same samples from species with low numbers of replicates (*F. zebrinus*, *F. nottii*, *F. sciadicus*).  


#dim(counts_design)

# -----------------------
# Determine which columns to drop
# based on sample size
# Drop columns with no data
# -----------------------

#length(counts_design)
samps <- as.data.frame(colnames(counts_design)[3:length(counts_design)])
samps <- samps %>% separate(`colnames(counts_design)[3:length(counts_design)]`, , into = c("species", "genus", "treatment","replicate"), sep = "_")
samps$group <- paste(samps$species,samps$genus,samps$treatment, sep = ":")
#table(samps$group)
#length(unique(paste(samps$species,samps$genus)))

samps_t <- samps %>% filter(treatment == "transfer")
samps_b <- samps %>% filter(treatment == "BW")
samps_f <- samps %>% filter(treatment == "FW")

#length(unique(paste(samps_t$species,samps_t$genus, samps_t$treatment)))
#length(unique(paste(samps_b$species,samps_b$genus, samps_b$treatment)))
#length(unique(paste(samps_f$species,samps_f$genus, samps_f$treatment)))
#setdiff(unique(paste(samps_f$species,samps_f$genus)),
#        unique(paste(samps_t$species,samps_t$genus)))

design <- counts_design[counts_design$Ensembl == 'Empty',]
#design$type <- c("species","native_salinity","clade","group","condition")

# dropping all zebrinus, nottii, sciadicus, due to sample size
drops <- c("X","Ensembl",
           "F_zebrinus_BW_1.quant","F_zebrinus_BW_2.quant",
           "F_zebrinus_FW_1.quant","F_zebrinus_FW_2.quant",
           "F_nottii_FW_1.quant","F_nottii_FW_2.quant",
           "F_sciadicus_BW_1.quant","F_sciadicus_FW_1.quant",
           "F_sciadicus_FW_2.quant","F_sciadicus_transfer_1.quant")

bw_drops <- grep('_BW_', colnames(counts_design), value=TRUE)

counts<-counts_design[!counts_design$Ensembl == 'Empty',]
rownames(counts)<-counts$Ensembl
design <- design[ , !(names(design) %in% drops)]
counts <- counts[ , !(names(counts) %in% drops)]
design <- design[ , !(names(design) %in% bw_drops)]
counts <- counts[ , !(names(counts) %in% bw_drops)]
print("dim design")
dim(design)
print("dim counts")
dim(counts)
gene.names<-rownames(counts)
design[] <- lapply( design, factor)



## Sample Design Matrix

#Same model with fixed effects, but condition = transfer: 

# ~physiology*condition*clade


# --------------------
# design categories
# --------------------

species<-as.character(unlist(design[1,]))
physiology<-as.character(unlist(design[2,]))
clade<-as.character(unlist(design[3,]))
condition<-as.character(unlist(design[5,]))
condition_physiology<-as.vector(paste(condition,physiology,sep="."))
condition_physiology_clade <- as.vector(paste(condition_physiology,clade,sep="."))
condition_physiology_clade <- as.vector(paste("group",condition_physiology_clade,sep=""))
cols<-colnames(counts)
ExpDesign <- data.frame(row.names=cols,
                        condition=condition,
                        physiology = physiology,
                        clade = clade,
                        species = species,
                        sample=cols)
#ExpDesign
# used for pairwise contrasts
#form<-as.formula("~0 + physiology*condition*clade")
form<-as.formula("~physiology*condition*clade")
design = model.matrix(form, ExpDesign)
#group <- interaction(physiology, condition, clade)
#mm <- model.matrix(~0 + group)
#colnames(design)
# check rank of matrix
#Matrix::rankMatrix( design )
#dim(design)
clade <- ExpDesign$clade
physiology <- ExpDesign$physiology

#write.table(ExpDesign,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/ExpDesign.txt",sep="\t",quote=FALSE, row.names = FALSE)


## Filtering and Normalization

# I kept all of this the same  


#Genes with low expression across samples were dropped from the analysis using a conservative approach. The function `filterByExpr` was used on the raw counts matrix. For each `condition_physiology` group (regardless of species), each sample must have a minium count of 10, and a group minimum total count of 100. This reduced the counts table to the following dimensions (genes x samples):

counts<-as.matrix(as.data.frame(sapply(counts, as.numeric)))
rownames(counts)<-gene.names
#class(counts)
#test<-counts %>% drop_na()
#test<-as.matrix(test)
lcom_unfilt<-log2(counts+1)
#plot(colSums(t(lcom_unfilt)))

keep<-filterByExpr(counts,design = design,group=condition_physiology,min.count = 10, min.total.count = 100)
counts.filt <- counts[keep,]
print("filtered_counts")
dim(counts.filt)
#write.table(counts.filt,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/exp.tsv",sep="\t",quote=FALSE)



## biomaRt annotation

# ============================================
# biomart annotation
# https://uswest.ensembl.org/Fundulus_heteroclitus/Info/Index
# ============================================

ensembl=useMart("ENSEMBL_MART_ENSEMBL")
ensembl = useDataset("fheteroclitus_gene_ensembl",mart=ensembl)
# test
# id <- c("ENSGMOP00000000001.1","ENSGMOP00000000002.1","ENSGMOP00000000003.1")
ensembl_proteinID <- rownames(counts)
ensembl_proteinID <- unlist(strsplit(ensembl_proteinID,split="[.]1"))

#length(ensembl_proteinID)
# can take >5 min
# do only once
ann<-getBM(attributes=c('ensembl_peptide_id','ensembl_transcript_id','ensembl_gene_id','gene_biotype','external_gene_name','description'), filters = 'ensembl_peptide_id', values = ensembl_proteinID, mart=ensembl)
#head(ann)
#dim(ann)
#length(unique(ann$ensembl_peptide_id))
ann <- ann[!duplicated(ann[,c(1)]),]
#dim(ann)
#write.table(ann,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/annotations.tsv",sep="\t",quote=FALSE)


## normalize


print("log counts before DE")
boxplot(log(counts.filt+1), las = 2, main = "")

genes = DGEList(count = counts.filt, group = condition_physiology_clade)
genes = calcNormFactors(genes)

# write normalized counts
tmp <- as.data.frame(cpm(genes))
tmp$Ensembl <- rownames(tmp)
tmp <- dplyr::select(tmp, Ensembl, everything())
#write.csv(tmp, file = file.path("C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/normalized_counts.csv"), quote = F, row.names = F)

# section 5.3 for normalizing for library size
#https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#normalising-gene-expression-distributions

```

# Normalize with Jane's method

## Read in library normalized counts
```{r, results="show",message=FALSE, warning=FALSE, error=TRUE}
#lib_norm_counts <- read.csv("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/blob/main/DE/DE_limma_FW_v_transfer/normalized_counts.csv?raw=true",stringsAsFactors = FALSE, header = TRUE,row.names = "Ensembl")
lib_norm_counts <- read.csv("C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/normalized_counts.csv",stringsAsFactors = FALSE, header = TRUE,row.names = "Ensembl")

head(lib_norm_counts)

```

## Normalize and Log Transform Read Counts

```{r}
#Transform count + constant:
constant <- 10
trans <- log2(lib_norm_counts+constant)
```

## Normalize log-transformed counts for all samples: 

```{r}
mean_counts <- colMeans(trans)
norm_counts <- sweep(trans, 2, colMeans(trans), "-")
grandmean <- mean(mean_counts)
norm_counts <- norm_counts + grandmean
head(norm_counts)
range(norm_counts)

```

## Calculate mean and standard deviation of each gene for every treatment after normalization, whole data set: 

```{r}
#Calculate and visualize mean and standard deviation of each gene for every treatment after normalization, whole data set: 
ttrans <- t(head(norm_counts))
ttrans <- t(norm_counts)
ttrans <- as.data.frame(ttrans)

meta <- read.table("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/blob/main/DE/ExpDesign_FW_v_transfer.txt?raw=true",stringsAsFactors = FALSE, header = TRUE)
rownames(meta) <- meta$sample
meta <- meta[,-5]
meta$physiology <- gsub('M', 'euryhaline', meta$physiology)
meta$physiology <- gsub('FW', 'stenohaline', meta$physiology)
meta$clade <- gsub('Clade1', 'clade 1', meta$clade)
meta$clade <- gsub('Clade2', 'clade 2', meta$clade)
meta$clade <- gsub('Clade3', 'clade 3', meta$clade)

	genelist <- colnames(ttrans)
	ttrans$trt <- meta$condition
	treat <- unique(ttrans$trt)

	#adjusting fun_MeanVar_all.R
	
	meandat <- matrix(nrow=length(treat), ncol=length(genelist))
	colnames(meandat) <- genelist
	rownames(meandat) <- treat 
	for (i in 1:length(treat)) {
		a <- which(ttrans$trt == treat[i])
		n <- apply(ttrans[a,1:(dim(ttrans)[2]-1)], 2, mean)
		b <- which(rownames(meandat)==treat[i])
		meandat[b,] <- n 
		}
	
	vardat <- matrix(nrow=length(treat), ncol=length(genelist))
	colnames(vardat) <- genelist
	rownames(vardat) <- treat 
	for (i in 1:length(treat)) {
		if (length(which(ttrans$trt==treat[i])) > 1) {
			a <- which(ttrans$trt==treat[i])
			#print(a)
			n <- apply(ttrans[a,1:(dim(ttrans)[2]-1)], 2, sd, na.rm=TRUE)
			#print(n)
			b <- which(rownames(vardat)==treat[i])
			#print(b)
			vardat[b,] <- n 
			}
	}
	mv <- list(mean=meandat, stdev=vardat)

#head(meanvar)
#dim(meanvar)

#mv <- MeanVar(ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)
slope <- (coef(lm(sd~mea))[2])
print(slope)


#Visualize mean-standard deviation relationship after normalization: 

#pdf('~/Documents/NIEHS_LSU_UCD/niehs/multipop_multigen/190924_normcountslog+15.pdf')
plot(mea,sd,xlab = "mean", pch=16, cex=0.5, main=paste("Mean-SD relationship of Normalized Log(counts + ",constant,"), slope=",round(slope,4),sep = ""))
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)
#dev.off()

#plot(colMeans(trans),sqrt(sd),xlab = "mean", pch=16, cex=0.5, main=paste("Mean-SD relationship of Normalized Log(counts + ",constant,"), slope=",round(slope,4),sep = ""))

```

# Use a strict count filter

```{r loadfiles, results='hide', include=FALSE, warning=FALSE, eval = FALSE}

min_count <- 30
min_total_count <- 300


#counts_design <- read.csv("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/raw/main/expression_tables/Ensembl_species_counts_designfactors.csv",stringsAsFactors = FALSE)
counts_design <- read.csv("C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/expression_tables/Ensembl_species_counts_designfactors.csv",stringsAsFactors = FALSE)


# I removed the same samples from species with low numbers of replicates (*F. zebrinus*, *F. nottii*, *F. sciadicus*).  


#dim(counts_design)

# -----------------------
# Determine which columns to drop
# based on sample size
# Drop columns with no data
# -----------------------

#length(counts_design)
samps <- as.data.frame(colnames(counts_design)[3:length(counts_design)])
samps <- samps %>% separate(`colnames(counts_design)[3:length(counts_design)]`, , into = c("species", "genus", "treatment","replicate"), sep = "_")
samps$group <- paste(samps$species,samps$genus,samps$treatment, sep = ":")
#table(samps$group)
#length(unique(paste(samps$species,samps$genus)))

samps_t <- samps %>% filter(treatment == "transfer")
samps_b <- samps %>% filter(treatment == "BW")
samps_f <- samps %>% filter(treatment == "FW")

#length(unique(paste(samps_t$species,samps_t$genus, samps_t$treatment)))
#length(unique(paste(samps_b$species,samps_b$genus, samps_b$treatment)))
#length(unique(paste(samps_f$species,samps_f$genus, samps_f$treatment)))
#setdiff(unique(paste(samps_f$species,samps_f$genus)),
#        unique(paste(samps_t$species,samps_t$genus)))

design <- counts_design[counts_design$Ensembl == 'Empty',]
#design$type <- c("species","native_salinity","clade","group","condition")

# dropping all zebrinus, nottii, sciadicus, due to sample size
drops <- c("X","Ensembl",
           "F_zebrinus_BW_1.quant","F_zebrinus_BW_2.quant",
           "F_zebrinus_FW_1.quant","F_zebrinus_FW_2.quant",
           "F_nottii_FW_1.quant","F_nottii_FW_2.quant",
           "F_sciadicus_BW_1.quant","F_sciadicus_FW_1.quant",
           "F_sciadicus_FW_2.quant","F_sciadicus_transfer_1.quant")

bw_drops <- grep('_BW_', colnames(counts_design), value=TRUE)

counts<-counts_design[!counts_design$Ensembl == 'Empty',]
rownames(counts)<-counts$Ensembl
design <- design[ , !(names(design) %in% drops)]
counts <- counts[ , !(names(counts) %in% drops)]
design <- design[ , !(names(design) %in% bw_drops)]
counts <- counts[ , !(names(counts) %in% bw_drops)]
print("dim design")
dim(design)
print("dim counts")
dim(counts)
gene.names<-rownames(counts)
design[] <- lapply( design, factor)



## Sample Design Matrix

#Same model with fixed effects, but condition = transfer: 

# ~physiology*condition*clade


# --------------------
# design categories
# --------------------

species<-as.character(unlist(design[1,]))
physiology<-as.character(unlist(design[2,]))
clade<-as.character(unlist(design[3,]))
condition<-as.character(unlist(design[5,]))
condition_physiology<-as.vector(paste(condition,physiology,sep="."))
condition_physiology_clade <- as.vector(paste(condition_physiology,clade,sep="."))
condition_physiology_clade <- as.vector(paste("group",condition_physiology_clade,sep=""))
cols<-colnames(counts)
ExpDesign <- data.frame(row.names=cols,
                        condition=condition,
                        physiology = physiology,
                        clade = clade,
                        species = species,
                        sample=cols)
#ExpDesign
# used for pairwise contrasts
#form<-as.formula("~0 + physiology*condition*clade")
form<-as.formula("~physiology*condition*clade")
design = model.matrix(form, ExpDesign)
#group <- interaction(physiology, condition, clade)
#mm <- model.matrix(~0 + group)
#colnames(design)
# check rank of matrix
#Matrix::rankMatrix( design )
#dim(design)
clade <- ExpDesign$clade
physiology <- ExpDesign$physiology

#write.table(ExpDesign,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/ExpDesign.txt",sep="\t",quote=FALSE, row.names = FALSE)


## Filtering and Normalization

# I kept all of this the same  


#Genes with low expression across samples were dropped from the analysis using a conservative approach. The function `filterByExpr` was used on the raw counts matrix. For each `condition_physiology` group (regardless of species), each sample must have a minium count of 10, and a group minimum total count of 100. This reduced the counts table to the following dimensions (genes x samples):

counts<-as.matrix(as.data.frame(sapply(counts, as.numeric)))
rownames(counts)<-gene.names
#class(counts)
#test<-counts %>% drop_na()
#test<-as.matrix(test)
lcom_unfilt<-log2(counts+1)
#plot(colSums(t(lcom_unfilt)))

keep<-filterByExpr(counts,design = design,group=condition_physiology,min.count = min_count, min.total.count = min_total_count)
counts.filt <- counts[keep,]
print("filtered_counts")
dim(counts.filt)
#write.table(counts.filt,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/exp.tsv",sep="\t",quote=FALSE)



## biomaRt annotation

# ============================================
# biomart annotation
# https://uswest.ensembl.org/Fundulus_heteroclitus/Info/Index
# ============================================

ensembl=useMart("ENSEMBL_MART_ENSEMBL")
ensembl = useDataset("fheteroclitus_gene_ensembl",mart=ensembl)
# test
# id <- c("ENSGMOP00000000001.1","ENSGMOP00000000002.1","ENSGMOP00000000003.1")
ensembl_proteinID <- rownames(counts)
ensembl_proteinID <- unlist(strsplit(ensembl_proteinID,split="[.]1"))

#length(ensembl_proteinID)
# can take >5 min
# do only once
ann<-getBM(attributes=c('ensembl_peptide_id','ensembl_transcript_id','ensembl_gene_id','gene_biotype','external_gene_name','description'), filters = 'ensembl_peptide_id', values = ensembl_proteinID, mart=ensembl)
#head(ann)
#dim(ann)
#length(unique(ann$ensembl_peptide_id))
ann <- ann[!duplicated(ann[,c(1)]),]
#dim(ann)
#write.table(ann,"C:/Users/jmcgirr/Documents/GitHub/RNAseq_17killifish_osmotic/DE/DE_limma_FW_v_transfer/annotations.tsv",sep="\t",quote=FALSE)

print("log counts before DE")
boxplot(log(counts.filt+1), las = 2, main = "")

genes = DGEList(count = counts.filt, group = condition_physiology_clade)
genes = calcNormFactors(genes)

head(genes)

# counts per million normalized by library size from genes$samples$norm.factors
cpm <- cpm(genes$counts, log = FALSE,normalized.lib.sizes = TRUE)

lib_norm_counts <- cpm
#lib_norm_counts$Ensembl <- NULL

head(lib_norm_counts)

#Transform count + constant:
constant <- 1
trans <- log2(lib_norm_counts+constant)

mean_counts <- colMeans(trans)
norm_counts <- sweep(trans, 2, colMeans(trans), "-")
grandmean <- mean(mean_counts)
norm_counts <- norm_counts + grandmean
head(norm_counts)
range(norm_counts)


ttrans <- t(head(norm_counts))
ttrans <- t(norm_counts)
ttrans <- as.data.frame(ttrans)

meta <- read.table("https://github.com/joemcgirr/RNAseq_17killifish_osmotic/blob/main/DE/ExpDesign_FW_v_transfer.txt?raw=true",stringsAsFactors = FALSE, header = TRUE)
rownames(meta) <- meta$sample
meta <- meta[,-5]
meta$physiology <- gsub('M', 'euryhaline', meta$physiology)
meta$physiology <- gsub('FW', 'stenohaline', meta$physiology)
meta$clade <- gsub('Clade1', 'clade 1', meta$clade)
meta$clade <- gsub('Clade2', 'clade 2', meta$clade)
meta$clade <- gsub('Clade3', 'clade 3', meta$clade)

	genelist <- colnames(ttrans)
	ttrans$trt <- meta$condition
	treat <- unique(ttrans$trt)

	#adjusting fun_MeanVar_all.R
	
	meandat <- matrix(nrow=length(treat), ncol=length(genelist))
	colnames(meandat) <- genelist
	rownames(meandat) <- treat 
	for (i in 1:length(treat)) {
		a <- which(ttrans$trt == treat[i])
		n <- apply(ttrans[a,1:(dim(ttrans)[2]-1)], 2, mean)
		b <- which(rownames(meandat)==treat[i])
		meandat[b,] <- n 
		}
	
	vardat <- matrix(nrow=length(treat), ncol=length(genelist))
	colnames(vardat) <- genelist
	rownames(vardat) <- treat 
	for (i in 1:length(treat)) {
		if (length(which(ttrans$trt==treat[i])) > 1) {
			a <- which(ttrans$trt==treat[i])
			#print(a)
			n <- apply(ttrans[a,1:(dim(ttrans)[2]-1)], 2, sd, na.rm=TRUE)
			#print(n)
			b <- which(rownames(vardat)==treat[i])
			#print(b)
			vardat[b,] <- n 
			}
	}
	mv <- list(mean=meandat, stdev=vardat)

#head(meanvar)
#dim(meanvar)

#mv <- MeanVar(ttrans)
mea <- as.vector(mv$mean)
sd <- as.vector(mv$stdev)
slope <- (coef(lm(sd~mea))[2])
print(slope)


#Visualize mean-standard deviation relationship after normalization: 

plot(mea,sd,xlab = "mean", pch=16, cex=0.5, main=paste("Mean-SD relationship of Normalized Log(counts + ",constant,"), slope=",round(slope,4),sep = ""))
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)

plot(mea,sqrt(sd),xlab = "mean", pch=16, cex=0.5, main=paste("Mean-SD relationship of Normalized Log(counts + ",constant,"), slope=",round(slope,4),sep = ""))
abline(coef(lm(sd~mea)),col="red",lwd=3)
lines(lowess(mea,sd,f=0.2),col="blue",lwd=3)

```


```{r}
####### prop DE/ME vs #########
library(evolqg)
library(ape)
library(geiger)
library(nlme)
library(phytools)
#install.packages("phangorn")
library(phangorn)
library(adephylo)
#install.packages("adephylo")

comps <-read.table("C:/Users/jmcgirr/Documents/all_2018_samples/prop_ME_CRPA_crosses.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
head(comps)
all_48 <- comps[which(comps$stage == "48hpf"),]
all_8 <- comps[which(comps$stage == "8dpf"),]
generalists_48 <- comps[which(comps$stage == "48hpf" & comps$specialist == "n"),]
generalists_8 <- comps[which(comps$stage == "8dpf" & comps$specialist == "n"),]

plot(comps$distance_km, comps$prop_DE)
plot(comps$distance_branchlength, comps$prop_DE)

#plot(generalists_48$distance_km, generalists_48$prop_DE)
plot(all_48$distance_km, all_48$prop_DE)
#plot(generalists_8$distance_km, generalists_8$prop_DE)
plot(all_8$distance_km, all_8$prop_DE)

plot(all_48$distance_branchlength, all_48$prop_DE)
plot(all_8$distance_branchlength, all_8$prop_DE)


all_cols <- c(dkr,dkr,lir,lir,yel,yel,pur,pur,bla,bla,blu,blu,red,red,whi,whi,gre,gre,grb,grb)
sep_cols <- c(dkr,lir,yel,pur,bla,blu,red,whi,gre,grb)
#plot(generalists_48$distance_branchlength, generalists_48$prop_DE)

par(mfrow=c(2,2))

comps <-read.table("C:/Users/jmcgirr/Documents/all_2018_samples/prop_DE_CRPA_crosses.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_48 <- comps[which(comps$stage == "48hpf"),]
all_8 <- comps[which(comps$stage == "8dpf"),]
gen_48 <- all_48[which(all_48$specialist == "n"),]
gen_8 <- all_8[which(all_8$specialist == "n"),]

#tiff("C:/Users/jmcgirr/Documents/all_2018_samples/manuscript_figs/de_me_48_8.tiff", width = 5, height = 5, units = 'in', res = 1000)
sep_cols <- c(dkr,lir,yel,pur,bla,blu,red,whi,gre,grb)
par(mfrow=c(2,2))
par(mai=c(0.2,0.4,0.2,0))
lm_all <- lm(all_48$distance_branchlength~all_48$prop_DE)
lm_gen <- lm(gen_48$distance_branchlength~gen_48$prop_DE)

plot(jitter(all_48$distance_branchlength), jitter((all_48$prop_DE)*100),pch = c(21,24)[as.factor(all_48$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,31),xlim = c(-0.02,0.45),xlab = "", ylab = "",xaxt ="n",cex.axis=1.5, yaxt = "n")
axis(side=2, at=c(0,10,20,30), labels = TRUE, cex.axis= 1.5)
abline(lm_all)
abline(lm_gen)

lm_all <- lm(all_8$distance_branchlength~all_8$prop_DE)
lm_gen <- lm(gen_8$distance_branchlength~gen_8$prop_DE)
par(mai=c(0.2,0.2,0.2,0.2))
plot(jitter(all_8$distance_branchlength), jitter((all_8$prop_DE)*100),pch = c(21,24)[as.factor(all_8$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,31),xlim = c(-0.02,0.45),xlab = "", ylab = "", yaxt = "n",xaxt ="n")
#abline(lm_all)
#abline(lm_gen)


comps <-read.table("C:/Users/jmcgirr/Documents/all_2018_samples/prop_ME_CRPA_crosses.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
head(comps)
all_48 <- comps[which(comps$stage == "48hpf" & comps$comp_type == "ph"),]
all_8 <- comps[which(comps$stage == "8dpf"& comps$comp_type == "ph"),]
sep_cols <- c(dkr,lir,yel,pur,grb)

par(mai=c(0.4,0.4,0,0))
plot(jitter(all_48$distance_branchlength), jitter((all_48$prop_DE)*100),pch = c(21,24)[as.factor(all_48$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,31),xlim = c(-0.02,0.45),xlab = "", ylab = "",cex.axis=1.5, yaxt = "n")
axis(side=2, at=c(0,10,20,30), labels = TRUE, cex.axis= 1.5)


par(mai=c(0.4,0.2,0,0.2))
plot(jitter(all_8$distance_branchlength), jitter((all_8$prop_DE)*100),pch = c(21,24)[as.factor(all_8$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,31),xlim = c(-0.02,0.45),xlab = "", ylab = "", yaxt = "n",cex.axis=1.5)
#dev.off()



## PGLS phylogenetic least squares ##https://www.r-phylo.org/wiki/HowTo/PGLS
## DE
all_48 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/de_v_crpa_48hpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_8 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/de_v_crpa_8dpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
row.names(all_48) <- all_48$tree_tip
row.names(all_8) <- all_8$tree_tip
all_48_de <- all_48
all_8_de <- all_8

# ME
all_48 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/mse_v_crpa_48hpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_8 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/mse_v_crpa_8dpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
tree <- read.tree("D:/Martin Lab/rna_2018/raxml/second_try_dna/RAxML_bestTree.58_dna_no_missing")
distances <- c(cophenetic(tree)["CAF1","CPF1"],cophenetic(tree)["CAF1","CMF1"],0,
               cophenetic(tree)["CAF1","CUNP"],cophenetic(tree)["CAF1","NAF1"])
all_48$distance_branch <- distances
all_8$distance_branch <- distances
row.names(all_48) <- all_48$tree_tip
row.names(all_8) <- all_8$tree_tip
all_48_me <- all_48
all_8_me <- all_8

all_48_de$prop_DE_vs_CRPA_scaled <- all_48_de$prop_DE_vs_CRPA *100
all_8_de$prop_DE_vs_CRPA_scaled <- all_8_de$prop_DE_vs_CRPA *100
all_48_me$prop_DE_vs_CRPA_scaled <- all_48_me$prop_DE_vs_CRPA *100
all_8_me$prop_DE_vs_CRPA_scaled <- all_8_me$prop_DE_vs_CRPA *100

## plot km on x axis
{

par(mfrow=c(2,2))

comps <-read.table("C:/Users/jmcgirr/Documents/all_2018_samples/prop_DE_CRPA_crosses.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_48 <- comps[which(comps$stage == "48hpf"),]
all_8 <- comps[which(comps$stage == "8dpf"),]
points_pch <- c(21,21,21,21,1,1,1,1,1,21)
sep_cols <- c("black","black",yel,blu,"black","black","black","black","black", red)

#tiff("C:/Users/jmcgirr/Documents/all_2018_samples/manuscript_figs/de_me_48_8_shapes.tiff", width = 5, height = 5, units = 'in', res = 1000)
par(mfrow=c(2,2))
par(mai=c(0.2,0.4,0.2,0))

plot(jitter(all_48$distance_km,50), all_48$prop_DE*100,col = "black",pch = points_pch,bg = sep_cols,
     cex = 2.3, ylim = c(0,31),xlab = "", ylab = "",xaxt ="n",cex.axis=1.5, yaxt = "n", xlim = c(-50,1200))
axis(side=2, at=c(0,10,20,30), labels = TRUE, cex.axis= 1.3)
pglsModel <- gls(prop_DE_vs_CRPA_scaled ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_48_de, method = "ML")
#summary(pglsModel)
abline(a = coef(pglsModel)[1], b = coef(pglsModel)[2], lty = 3)

par(mai=c(0.2,0.2,0.2,0.2))
plot(jitter(all_8$distance_km,50), all_8$prop_DE*100,col = "black",pch = points_pch,bg = sep_cols,
     cex = 2.3, ylim = c(0,31),xlab = "", ylab = "",xaxt ="n",cex.axis=1.5, yaxt = "n", xlim = c(-50,1200))
pglsModel <- gls(prop_DE_vs_CRPA_scaled ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_8_de, method = "ML")
#summary(pglsModel)
#abline(a = coef(pglsModel)[1], b = coef(pglsModel)[2], lty = 3)

comps <-read.table("C:/Users/jmcgirr/Documents/all_2018_samples/prop_ME_CRPA_crosses.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
head(comps)
all_48 <- comps[which(comps$stage == "48hpf"),]
all_8 <- comps[which(comps$stage == "8dpf"),]
sep_cols <- c("grey","grey",gre,grb,"grey","grey","grey","grey","grey", lir)

par(mai=c(0.4,0.4,0,0))
plot(jitter(all_48$distance_km,50), all_48$prop_DE*100,col = "black",pch = points_pch,bg = sep_cols,
     cex = 2.3, ylim = c(0,31),xlab = "", ylab = "",cex.axis=1.5, yaxt = "n", xlim = c(-50,1200), xaxt = "n")
axis(side=2, at=c(0,10,20,30), labels = TRUE, cex.axis= 1.3)
axis(side=1, at=c(0,250,500,750,1000), labels = c(0,250,500,750,1000), cex.axis= 1)

pglsModel <- gls(prop_DE_vs_CRPA_scaled ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_48_me, method = "ML")
#summary(pglsModel)
#abline(a = coef(pglsModel)[1], b = coef(pglsModel)[2], lty = 3)

par(mai=c(0.4,0.2,0,0.2))
plot(jitter(all_8$distance_km,50), all_8$prop_DE*100,col = "black",pch = points_pch,bg = sep_cols,
     cex = 2.3, ylim = c(0,31),xlab = "", ylab = "",cex.axis=1.5, yaxt = "n", xlim = c(-50,1200), xaxt = "n")
axis(side=1, at=c(0,250,500,750,1000), labels = c(0,250,500,750,1000), cex.axis= 1)
pglsModel <- gls(prop_DE_vs_CRPA_scaled ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_8_me, method = "ML")
#summary(pglsModel)
#abline(a = coef(pglsModel)[1], b = coef(pglsModel)[2], lty = 3)
#dev.off()
}





generalists_48 <- comps[which(comps$stage == "48hpf" & comps$specialist == "n"),]
generalists_8 <- comps[which(comps$stage == "8dpf" & comps$specialist == "n"),]

plot(comps$distance_km, comps$prop_DE)
plot(comps$distance_branchlength, comps$prop_DE)

#plot(generalists_48$distance_km, generalists_48$prop_DE)
plot(all_48$distance_km, all_48$prop_DE)
#plot(generalists_8$distance_km, generalists_8$prop_DE)
plot(all_8$distance_km, all_8$prop_DE)


#plot(generalists_48$distance_branchlength, generalists_48$prop_DE)
#tiff("C:/Users/jmcgirr/Documents/all_2018_samples/manuscript_figs/de_48.tiff", width = 4.5, height = 5, units = 'in', res = 1000)
plot(jitter(all_48$distance_branchlength), jitter((all_48$prop_DE)*100),pch = c(21,24)[as.factor(all_48$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,21.5),xlim = c(-0.02,0.45),xlab = "", ylab = "",xaxt ="n",cex.axis=1.5)
#dev.off()
#plot(generalists_8$distance_branchlength, generalists_8$prop_DE)
#tiff("C:/Users/jmcgirr/Documents/all_2018_samples/manuscript_figs/de_8.tiff", width = 4.5, height = 5, units = 'in', res = 1000)
plot(jitter(all_8$distance_branchlength), jitter((all_8$prop_DE)*100),pch = c(21,24)[as.factor(all_8$point_shape)],
     bg =sep_cols,col = "black", cex = 2.3, ylim = c(0,21.5),xlim = c(-0.02,0.45),xlab = "", ylab = "", yaxt = "n",xaxt ="n")
#dev.off()



## PGLS phylogenetic least squares ##https://www.r-phylo.org/wiki/HowTo/PGLS
# ME
all_48 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/mse_v_crpa_48hpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_8 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/mse_v_crpa_8dpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
tree <- read.tree("D:/Martin Lab/rna_2018/raxml/second_try_dna/RAxML_bestTree.58_dna_no_missing")
distances <- c(cophenetic(tree)["CAF1","CPF1"],cophenetic(tree)["CAF1","CMF1"],0,
               cophenetic(tree)["CAF1","CUNP"],cophenetic(tree)["CAF1","NAF1"])
all_48$distance_branch <- distances
all_8$distance_branch <- distances
species<-c("CAF1",	"CMF1",	"CPF1",	"CUNP",	"NAF1")
all_tree<-drop.tip(tree,tree$tip.label[-match(species, tree$tip.label)])
species<-c("CAF1",	"CUNP",	"NAF1")
generalist_tree<-drop.tip(tree,tree$tip.label[-match(species, tree$tip.label)])
generalists_48 <- all_48[which(all_48$specialist == "n"),]
generalists_8 <- all_8[which(all_8$specialist == "n"),]
test<-c("CAF1",	"CUNP",	"NAF1")
row.names(all_48) <- all_48$tree_tip
row.names(all_8) <- all_8$tree_tip
row.names(generalists_48) <- test
row.names(generalists_8) <- test
all_48 <- all_48_me
all_8 <- all_8_me

pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = generalist_tree),
                 data = generalists_48, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = generalist_tree),
                 data = generalists_8, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_48, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_8, method = "ML")
summary(pglsModel)

## DE
all_48 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/de_v_crpa_48hpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
all_8 <- read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/de_v_crpa_8dpf.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
generalists_48 <- all_48[which(all_48$specialist == "n"),]
generalists_8 <- all_8[which(all_8$specialist == "n"),]
row.names(all_48) <- all_48$tree_tip
row.names(all_8) <- all_8$tree_tip
test<-c("NAF1",	"CUNP",	"CAF1")
row.names(generalists_48) <- test
row.names(generalists_8) <- test
all_48 <- all_48_me
all_8 <- all_8_me

pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = generalist_tree),
                 data = generalists_48, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = generalist_tree),
                 data = generalists_8, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_48, method = "ML")
summary(pglsModel)
pglsModel <- gls(prop_DE_vs_CRPA ~ distance_km, correlation = corBrownian(phy = all_tree),
                 data = all_8, method = "ML")
summary(pglsModel)

#plot tree 
species<-c("CAF1",	"CMF1",	"CPF1",	"CUNP",	"NAF1", "ETAA1","OAF1","OMF1","OPF1")
map_tree<-drop.tip(tree,tree$tip.label[-match(species, tree$tip.label)])
map_tree<-rotate(map_tree,15)
nodelabels()
#tiff("D:/Martin Lab/rna_2018/raxml/map_tree_cp_and_op.tiff", width = 4.5, height = 4.5, units = 'in', res = 1000)
#png("D:/Martin Lab/rna_2018/raxml/map_tree_cp_and_op.png", width = 5, height = 4.5, units = 'in', res = 1000,bg = "transparent")
plotTree(map_tree,offset=1,direction = "leftwards")
nodelabels(c("",100,100,100,100,100,100,100),frame="none",adj=c(1.2,0),cex = 0.4 )
add.scale.bar(0.6,7)
#dev.off()

#png("D:/Martin Lab/rna_2018/raxml/map_tree_cp_and_op.png", width = 6, height = 10, units = 'in', res = 1000,bg = "transparent")
plotTree(tree,offset=1)
#dev.off()



# phylogenetic mantel test 
tree <- read.tree("D:/Martin Lab/rna_2018/raxml/second_try_dna/RAxML_bestTree.58_dna_no_missing")
distance_km <-as.matrix(read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/distance_km_matrix.txt" ,header = TRUE,row.names=1))
mis_48 <-as.matrix(read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/misexpression_48hpf_matrix.txt" ,header = TRUE,row.names=1))
mis_8 <-as.matrix(read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/misexpression_8dpf_matrix.txt" ,header = TRUE,row.names=1))
#mis_48 <-as.matrix(read.table("C:/Users/jmcgirr/Documents/all_2018_samples/mse_v_distance/na_rm_test_matrix.txt" ,header = TRUE,row.names=1))

MyTree <- read.tree("D:/Martin Lab/rna_2018/raxml/second_try_dna/RAxML_bestTree.58_dna_no_missing")
head(MyTree)

tree = MyTree
species<-c("CAF1",	"CMF1",	"CPF1",	"CUNP",	"NAF1")
tree<-drop.tip(tree,tree$tip.label[-match(species, tree$tip.label)])

PhyloMantel(tree, mis_48, distance_km, k = 10000)
MantelCor(mis_48, distance_km)
## End(Not run)


#####

```
